### PR TITLE
Metrics from VICI client

### PR DESCRIPTION
**Pull request by Crevil**
This change introduces prometheus metrics from a strongswan process.
It collects statistics by inspecting the results of `swanctl --list-conn` and `swanctl --list-sas` calls.

Package `vici` is ported from https://github.com/bronze1man/goStrongswanVici and updated with error handling and removal of `fmt.Print` calls.

Package `strongswan` implements the actually collection of data and reports to the `PrometheusReporter` instance.

### CODE REVIEW

**Comment by eXeDK** [comment]
This is pretty cool 👍 💯

File: internal/vici/monitorSA.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+		if err != nil {
+			return fmt.Errorf("convert from general: %w", err)
+		}
+		event.Ike[name] = sa
+	}
+	callback(eventName, event)
+	return nil
+}
+
+func (c *ClientConn) MonitorSA(callback MonitorCallBack, watchdog time.Duration) (err error) {
```
Comment (Crevil) [inline internal/vici/monitorSA.go]: This looks great in terms of tracking up-down changes between our daemon checks. I'd like to get this change in sooner that later and then make a separate PR using this for event better metrics.

File: internal/metrics/metrics.go [hunk 2]
```diff
@@ -73,17 +103,121 @@ func NewPrometheusReporter(reg prometheus.Registerer) (*PrometheusReporter, erro
...
+				Subsystem: subSystemIKE,
+				Name:      "packets_out_total",
+				Help:      "Total number of transmitted packets",
+			}, []string{childSAName}),
+			lastPacketInSeconds: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subSystemIKE,
+				Name:      "packets_in_silence_duration_seconds",
+				Help:      "Duration of silences between packets in",
+				Buckets:   prometheus.ExponentialBuckets(0.1, 10, 10),
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/metrics/metrics.go]: Tried to write out these buckets, and they seem a bit off. I would go with 
```prometheus.ExponentialBuckets(15, 2, 14)```
that will bucket it in reasonable time frames from 15 seconds to 1,4 day

File: internal/metrics/metrics.go [hunk 3]
```diff
@@ -73,17 +103,121 @@ func NewPrometheusReporter(reg prometheus.Registerer) (*PrometheusReporter, erro
...
+			}, []string{}),
+			lastPacketOutSeconds: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subSystemIKE,
+				Name:      "packets_out_silence_duration_seconds",
+				Help:      "Duration of silences between packets out",
+				Buckets:   prometheus.ExponentialBuckets(0.1, 10, 10),
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/metrics/metrics.go]: Should be the  same as `lastPacketInSeconds`

File: internal/metrics/metrics.go [hunk 4]
```diff
@@ -73,17 +103,121 @@ func NewPrometheusReporter(reg prometheus.Registerer) (*PrometheusReporter, erro
...
+				Subsystem: subSystemIKE,
+				Name:      "installs_total",
+				Help:      "Total number of SA installs",
+			}, []string{}),
+			rekeySeconds: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subSystemIKE,
+				Name:      "rekey_seconds",
+				Help:      "Duration between re-keying",
+				Buckets:   []float64{10, 30, 60, 120, 300, 480, 600},
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/metrics/metrics.go]: Think we should add a few higher buckets. We could use the one for `lastPacketInSeconds` with a bit lower count: 
```
prometheus.ExponentialBuckets(15, 2, 12)
```

File: internal/metrics/metrics.go [hunk 5]
```diff
@@ -73,17 +103,121 @@ func NewPrometheusReporter(reg prometheus.Registerer) (*PrometheusReporter, erro
...
+			}, []string{}),
+			lifeTimeSeconds: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subSystemIKE,
+				Name:      "lifetime_seconds",
+				Help:      "Duration of each IKE session",
+				Buckets:   prometheus.ExponentialBuckets(0.1, 10800, 10), // 3 hours
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/metrics/metrics.go]: I think we should go with the 
```prometheus.ExponentialBuckets(15, 2, 14)```

File: internal/metrics/metrics.go [hunk 6]
```diff
@@ -101,3 +235,124 @@ func (r *tcpChecker) ReportPortCheck(report tcpchecker.Report) {
 	}
 	r.previousOpenState = &report.Open
 }
+
+func (p *PrometheusReporter) IKESAStatus(conn vici.IKEConf, sa *vici.IkeSa) {
+	if sa == nil {
+		p.logger.Errorf("No SA for connecetion configuration: %#v", conn)
```
[the remark this answers is not in the document]
Comment (emilingerslev) [reply internal/metrics/metrics.go]: me to...

File: internal/metrics/metrics.go [hunk 7]
```diff
@@ -101,3 +235,123 @@ func (r *tcpChecker) ReportPortCheck(report tcpchecker.Report) {
...
+func (p *PrometheusReporter) setHistogramByMax(h *prometheus.HistogramVec, value, name string) {
+	max, ok := p.maxValue(name, value)
+	if !ok {
+		return
+	}
+	h.WithLabelValues().Observe(max)
+}
+
+func (p *PrometheusReporter) setHistogramByMin(h *prometheus.HistogramVec, value, name string) {
+	max, ok := p.minValue(name, value)
```
Comment (emilingerslev) [inline internal/metrics/metrics.go]: copy paste error? :)

**Comment by Crevil** [comment]
I've not set any of all the labels we discussed @emilingerslev . And I've removed the notes we made when discussing it, so maybe we should give it another try?
Currently I only have child SA names on the metrics relevant for that.

**Comment by emilingerslev** [comment]
Alright, so yeah maybe we should go over the labels again. From the top of my head I got for Child SA's
* Local IP range
* Remote IP range

For IKE SA's (and the Child SA's below):
* Local Peer IP
* Remote Peer IP

**Comment by Crevil** [comment]
I'll add that. Then we can get it merged and out and running. Then we can add new labels later on.

**Comment by kaspernissen** [comment]
atlantis plan

**Comment by HamAstrochimp** [comment]
Ran Plan for 0 projects:

**Comment by Crevil** [comment]
I've added labels. An example metric is now like this.

```
strong_duckling_ike_sa_bytes_in_total{child_sa_name="net-net-0",ike_sa_name="gw-gw",instance="strongswan1:8000",job="strongswan1",local_ip_range="10.101.0.1/32",local_peer_ip="172.22.0.2",remote_ip_range="10.102.0.1/32",remote_peer_ip="172.22.0.3"}
```

**Comment by Crevil** [comment]
@emilingerslev ready for another review of the changes.

**Comment by Crevil** [comment]
Great. I'll work on a good commit message :D

**Pull request merged**